src/main/scala/is/hail/methods/MendelErrors.scala

```diff
@@ -74,47 +70,36 @@ object MendelErrors {
     if (nSamplesDiscarded > 0)
       warn(s"$nSamplesDiscarded ${ plural(nSamplesDiscarded, "sample") } discarded from .fam: sex of child is missing.")
 
-    val sampleTrioRoles = mutable.Map.empty[String, List[(Int, Int)]]
-    trios.zipWithIndex.foreach { case (t, ti) =>
-      sampleTrioRoles += (t.kid -> sampleTrioRoles.getOrElse(t.kid, List.empty[(Int, Int)]).::(ti, 0))
-      sampleTrioRoles += (t.knownDad -> sampleTrioRoles.getOrElse(t.knownDad, List.empty[(Int, Int)]).::(ti, 1))
-      sampleTrioRoles += (t.knownMom -> sampleTrioRoles.getOrElse(t.knownMom, List.empty[(Int, Int)]).::(ti, 2))
-    }
+    val trioMatrix = vds.trioMatrix(Pedigree(trios), completeTrios = true)
+    val rowType = trioMatrix.rowType
+    val nTrios = trioMatrix.nSamples
 
     val sc = vds.sparkContext
-    val sampleTrioRolesBc = sc.broadcast(sampleTrioRoles)
     val triosBc = sc.broadcast(trios)
     // all trios have defined sex, see filter above
     val trioSexBc = sc.broadcast(trios.map(_.sex.get))
 
-    val zeroVal: MultiArray2[GenotypeType] = MultiArray2.fill(trios.length, 3)(NoCall)
-
-    def seqOp(a: MultiArray2[GenotypeType], s: Annotation, g: Genotype): MultiArray2[GenotypeType] = {
-      sampleTrioRolesBc.value.get(s.asInstanceOf[String]).foreach(l => l.foreach { case (ti, ri) => a.update(ti, ri, Genotype.gtType(g)) })
-      a
-    }
-
-    def mergeOp(a: MultiArray2[GenotypeType], b: MultiArray2[GenotypeType]): MultiArray2[GenotypeType] = {
-      for ((i, j) <- a.indices)
-        if (b(i, j) != NoCall)
-          a(i, j) = b(i, j)
-      a
-    }
-
     new MendelErrors(vds.hc, vds.vSignature, trios, vds.stringSampleIds,
-      vds
-        .aggregateByVariantWithKeys(zeroVal)(
-          (a, v, s, g) => seqOp(a, s, g),
-          mergeOp)
-        .flatMap { case (v, a) =>
-          a.rows.flatMap { case (row) => val code = getCode(row, v.copyState(trioSexBc.value(row.i)))
+      trioMatrix.rdd2.mapPartitions { it =>
+        val view = new HardcallTrioGenotypeView(rowType, "gt")
+        it.flatMap { rv =>
+          view.setRegion(rv)
+          val v = Variant.fromRegionValue(rv.region, rowType.loadField(rv, 1))
+          Range(0, nTrios).flatMap { i =>
```

Comment:
We usually write this as `(0 until nTrios)`.

I kind of feel like this should be:

```scala
(0 until nTrios).iterator.flatMap { i => ...
```

so as to avoid allocating an `IndexedSeq` that will just get converted into an `Iterator` by `Iterator.flatMap`.

Reply:
Good call.

Reply:
`Iterator.range(0, nTrios)` saves even allocating the `0 until nTrios` Range object. Do you think it's worth the slight hit to readability?

Reply:
That looks good to me.